File: sorting/csharp/SelectionSort.cs
namespace Sorting.Algorithms
{
    /// <summary>
    /// Selection sort algorithm
    ///  Θ(n) ~ N^2
    /// </summary>
    public class SelectionSort : ISortingAlgorithm
    {
        public int[] sort(int[] array)
        {
            // for all elements in array
            for(int i = 0; i <= array.Length; i++)
            {
                // assume that element under cursor element is minimal 
                // (everything left should be already sorted)
                int min = i; 

                // for all following elements after cursor (find the minimal in sub array)
                for(int j = i; j < array.Length; j++)
                {
                    // if element on 'j' position is less that element on 'min' position
                    if(array[min] > array[j])
                        // change reference to 'min' position
                        min = j;
                }

                // if current element 'i' is not minimal
                if(min != i)
                {
                    // change value of current element and minimal element
                    int temp = array[i];
                    array[i] = array[min];
                    array[min] = temp;
                }       
            }
            // array should be completely sorted here!
            return array;
        }
    }
}

File: sorting/SelectionSort.java

/**
 * @description Selection sort algorithm implementation
 * @author Yuriy Shapovalov
 */
public class SelectionSort {

	public static <T extends Comparable<T>> void sort(T[] arr) {
		// assume empty or one-element array is sorted by definition
		if (arr.length <= 1) {
			return;
		}
		// for all elements in array
		for(int i = 0; i < arr.length; i++) {
			// assume that current element is minimal
			int min = i;
			
			// iterate all elements from 'i' to the end of array (to the right) 
			// and find element smaller than current
			for(int j = i; j <= arr.length; j++) {
				if(arr[min].compareTo(arr[j]) > 0) {
					min = j;
				}
			}
			
			// if found element smaller than 'i', put it into the place of 'i'
			if(min != i)
			{
				T temp = arr[i];
				arr[i] = arr[min];
				arr[min] = arr[i];
			}
		}
	}
}